pennylane_lightning/core/src/simulators/lightning_kokkos/StateVectorKokkos.hpp

```diff
@@ -757,6 +757,60 @@ class StateVectorKokkos final
         return -static_cast<fp_t>(0.5);
     }
 
+    /**
+     * @brief Collapse the state vector as after having measured one of the
+     * qubits.
+     *
+     * The branch parameter imposes the measurement result on the given wire.
+     *
+     * @param wire Wire to collapse.
+     * @param branch Branch 0 or 1.
+     */
+    void collapse(const std::size_t wire, const bool branch) {
+        auto &&num_qubits = this->getNumQubits();
+
+        const size_t stride = pow(2, num_qubits_ - (1 + wire));
+        const size_t vec_size = pow(2, num_qubits_);
+        const auto section_size = vec_size / stride;
+        const auto half_section_size = section_size / 2;
+
+        const size_t negbranch = branch ? 0 : 1;
+
+        Kokkos::MDRangePolicy<DoubleLoopRank> policy_2d(
+            {0, 0}, {half_section_size, stride});
+        Kokkos::parallel_for(
+            policy_2d,
+            collapseFunctor<fp_t>(*data_, num_qubits, stride, negbranch));
+
+        normalize();
+    }
+
+    /**
+     * @brief Normalize vector (to have norm 1).
+     */
+    void normalize() {
+        KokkosVector sv_view =
+            getView(); // circumvent error capturing this with KOKKOS_LAMBDA
+
+        // TODO: @tomlqc what about squaredNorm()
+        PrecisionT squaredNorm = 0.0;
+        Kokkos::parallel_reduce(
+            sv_view.size(),
+            KOKKOS_LAMBDA(const size_t i, PrecisionT &sum) {
+                sum += std::norm<PrecisionT>(sv_view(i));
```

Comment:
`std::norm` is a host function, so it won't work on GPUs for example. We should use the Kokkos types' methods within kernels.
```suggestion
                sum += norm(sv_view(i));
```